lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,961 @@
+    if (getaddrinfo(hostname, servname, &hints, &res) != 0) {
+        goto Error;
+    }
+
+    switch(res->ai_family) {
+        case AF_INET:
+        case AF_INET6:
+            memcpy(ss, res->ai_addr, res->ai_addrlen);
+            break;
+        default:
+            goto Error;
+    }
+
+    socklen_t len =  res->ai_addrlen;
+    freeaddrinfo(res);
+    return len;
+
+Error:
+    if (res != NULL)
+        freeaddrinfo(res);
+    return 0;
+}
+
+static socklen_t get_sockname(h2o_conn_t *_conn, struct sockaddr *sa)
+{
+    struct st_mruby_subreq_conn_t *conn = (void *)_conn;
+    if (conn->server.host.base != NULL) {
+        struct st_mruby_subreq_t *subreq = H2O_STRUCT_FROM_MEMBER(struct st_mruby_subreq_t, conn, conn);
+        conn->server.len = parse_hostport(&subreq->super.pool, conn->server.host, conn->server.port, &conn->server.addr);
+        conn->server.host.base = NULL;
+    }
+    memcpy(sa, &conn->server.addr, conn->server.len);
+    return conn->server.len;
+}
+
+static socklen_t get_peername(h2o_conn_t *_conn, struct sockaddr *sa)
+{
+    struct st_mruby_subreq_conn_t *conn = (void *)_conn;
+    if (conn->remote.host.base != NULL) {
+        struct st_mruby_subreq_t *subreq = H2O_STRUCT_FROM_MEMBER(struct st_mruby_subreq_t, conn, conn);
+        conn->remote.len = parse_hostport(&subreq->super.pool, conn->remote.host, conn->remote.port, &conn->remote.addr);
+        conn->remote.host.base = NULL;
+    }
+    memcpy(sa, &conn->remote.addr, conn->remote.len);
+    return conn->remote.len;
+}
+
+static h2o_socket_t *get_socket(h2o_conn_t *conn)
+{
+    return NULL;
+}
+
+static int handle_header_env_key(h2o_mruby_shared_context_t *shared_ctx, h2o_iovec_t *env_key, h2o_iovec_t value, void *_req)
+{
+    h2o_req_t *req = _req;
+    const h2o_token_t *token;
+
+    /* convert env key to header name (lower case) */
+    h2o_iovec_t name = convert_env_to_header_name(&req->pool, env_key->base, env_key->len);
+    if (name.base == NULL)
+        return 0;
+
+    if ((token = h2o_lookup_token(name.base, name.len)) != NULL) {
+        if (token == H2O_TOKEN_CONTENT_LENGTH) {
+            /* skip. use CONTENT_LENGTH instead of HTTP_CONTENT_LENGTH */
+        } else {
+            value = h2o_strdup(&req->pool, value.base, value.len);
+            h2o_add_header(&req->pool, &req->headers, token, NULL, value.base, value.len);
+        }
+    } else {
+        value = h2o_strdup(&req->pool, value.base, value.len);
+        h2o_add_header_by_str(&req->pool, &req->headers, name.base, name.len, 0, NULL, value.base, value.len);
+    }
+
+    return 0;
+}
+
+static void on_subreq_error_callback(void *data, h2o_iovec_t prefix, h2o_iovec_t msg)
+{
+    struct st_mruby_subreq_t *subreq = (void *)data;
+    mrb_state *mrb = subreq->ctx->shared->mrb;
+
+    assert(!mrb_nil_p(subreq->error_stream));
+
+    h2o_iovec_t list[] = {prefix, msg};
+    h2o_iovec_t concat = h2o_concat_list(&subreq->super.pool, list, 2);
```

Comment:
Please use `h2o_concat(&subreq->super.pool, prefix, msg)` to avoid constructing a temporary object (i.e. `list[]).